### DISCUSSION TITLE
What are the benefits of lazy computation?

### DISCUSSION THREAD

**Discussion by douglasdew** [ml-explore/mlx#387] [General]
Also, is there a flag of some kind which can be set to disable lazy computation? When in learning mode (as I am) and stepping through the debugger while watching array values, it might be better to have arrays evaluated immediately.

**Comment by awni** [comment]
> Also, is there a flag of some kind which can be set to disable lazy computation? When in learning mode (as I am) and stepping through 

Lazy computation is one way to let us get the graph for transformations (grad/vmap/optimizations) without doing wasteful work.

> When in learning mode (as I am) and stepping through the debugger while watching array values, it might be better to have arrays evaluated immediately.

I very much hear you. For Python this works fine since if you inspect an array it will eval it (and any higher level bindings we intend to do the same). From GDB w/ C++ it is not so clean. But basically you need to call the print function on the array to look at it's contents.

**Reply by douglasdew** [reply]
As always, thank you very much for being helpful @awni.

Although not a python developer, I decided to adjust my c++ course by first familiarizing myself with the python code. So, I'm stepping through the python examples now. Along the way, I'm developing a grudging envy for the array indexing/slicing functionality which python offers. It makes sense to me that you and other data scientists would make use of python.

**Reply by awni** [reply]
> Along the way, I'm developing a grudging envy for the array indexing/slicing functionality which python offers. 

We are planning to add better support for this in C++ (and of course it should be supported in swift as well)

**Reply by douglasdew** [reply]
One of the possibilities which I most enjoy about this MLX effort is that because this effort is Apple-platform-focused, the swift compiler team might take an interest and add language features so that swift is just as good as python for array indexing/slicing.
